### PR TITLE
feat: Add push event controller

### PR DESCRIPTION
**Pull request by mimotej**
Resolves: #10 
Handles push event send to controller. Creates a TaskRun which runs Task specified in #20.
~~Still needs tests to be implemented.~~ I am gonna split this and do the tests later down the line.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -72,6 +72,44 @@ export = (app: Probot) => {
     };
     await k8sApi.createNamespacedSecret(namespace,secret);
   });
+
+  //handle push event
+  app.on("push", async (context) => {
+    kc.loadFromDefault();
+    const client = kc.makeApiClient(k8s.CustomObjectsApi);
```
Comment (tumido) [inline src/index.ts:79]: Please create this client at global scope

File: src/index.ts [hunk 2]
```diff
@@ -72,6 +72,44 @@ export = (app: Probot) => {
     };
     await k8sApi.createNamespacedSecret(namespace,secret);
   });
+
+  //handle push event
+  app.on("push", async (context) => {
+    kc.loadFromDefault();
```
Comment (tumido) [inline src/index.ts:78]: Why is this needed here?

**Review by tumido — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 1 above)
Comment (mimotej) [reply src/index.ts]: Resolved in latest commit.

File: src/index.ts [hunk 3]
```diff
@@ -72,6 +72,44 @@ export = (app: Probot) => {
...
+    try {
+      const taskRun = await client.createNamespacedCustomObject(
+        "tekton.dev",
+        "v1beta1",
+        namespace,
+        "taskruns",
+        payload
+      )
+    }catch(e){
+      console.log(e)
```
[the remark this answers is not in the document]
Comment (mimotej) [reply src/index.ts]: Resolved in latest commit.

File: src/index.ts [hunk 4]
```diff
@@ -72,6 +72,44 @@ export = (app: Probot) => {
     };
     await k8sApi.createNamespacedSecret(namespace,secret);
   });
+
+  //handle push event
+  app.on("push", async (context) => {
+    kc.loadFromDefault();
+    const client = kc.makeApiClient(k8s.CustomObjectsApi);
+    var configExists = Boolean(context.payload.commits?.reduce((acc, commit) => (acc = [], acc.push(...commit.added), acc.push(...commit.modified), acc.push(...commit.deleted), acc.find((name:String) => name == "peribolos.yaml")), {}));
```
[the remark this answers is not in the document]
Comment (mimotej) [reply src/index.ts]: Thanks for ideas :+1:, wasn't 100 % sure how this works, but I sat down and looked into it a little bit and changed it in the latest commit. Now I should be accumulating results and I am running `find()` on the result. WDYT @tumido

File: src/index.ts [hunk 5]
```diff
@@ -72,6 +73,43 @@ export = (app: Probot) => {
     };
     await k8sApi.createNamespacedSecret(namespace,secret);
   });
+
+  //handle push event
+  app.on("push", async (context) => {
+
+    const configExists = Boolean(context.payload.commits?.reduce((acc, commit) => acc = [...acc, ...commit.added, ...commit.modified,...commit.removed], []).find((name:String) => name == "peribolos.yaml"));
```
Comment (tumido) [inline src/index.ts:80]: While this seems to work, I don't think it's correct. Reduce accepts a callback consuming an accumulator and current value. This callback has a body and it's return value is used as accumulator. Instead of a function body you're doing an assignment.. and since your assignment is not a statement but an expression, it also returns the assigned value:

```
a = 5
// returns 5
// vs.
var a = 5
// returns undefined
```

Hence what you're doing here is a bit of an abuse and you should return new accumulator value only.

It should be:
```suggestion
    const configExists = Boolean(context.payload.commits?.reduce((acc, commit) => [...acc, ...commit.added, ...commit.modified,...commit.removed], []).find((name: String) => name == "peribolos.yaml"));
```

**Review by tumido — CHANGES_REQUESTED** [review]
1 small nit, otherwise looks great!

File: src/index.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ const kc = new k8s.KubeConfig();
 kc.loadFromDefault();
 const k8sApi = kc.makeApiClient(k8s.CoreV1Api);
 const k8sContext = kc.getCurrentContext();
+const client = kc.makeApiClient(k8s.CustomObjectsApi);
```
Comment (harshad16) [inline src/index.ts:9]: should we make this consistent with other k8s API object like this
https://github.com/open-services-group/peribolos-as-a-service/pull/41/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R8 ?

just for easy of readability

File: src/index.ts [hunk 7]
```diff
@@ -72,6 +73,43 @@ export = (app: Probot) => {
...
+    try {
+      await client.createNamespacedCustomObject(
+        "tekton.dev",
+        "v1beta1",
+        namespace,
+        "taskruns",
+        payload
+      )
+    }catch(e: any){
+      app.log.error(e.message);
```
Comment (harshad16) [inline src/index.ts:110]: asking from a perspective of nodejs rookie :sweat_smile: 
any specific reason of using try block here , instead of the `then` method of function?

**Review by harshad16 — COMMENTED** [review]
lgtm
just few question for my better understanding of nodejs code base.

File: src/index.ts (hunk 6 above)
Comment (mimotej) [reply src/index.ts]: Good point changed it. Thanks :+1:

**Review by tumido — COMMENTED** [review]
Few logging nits, otherwise lgtm :+1: Good job!

File: src/index.ts [hunk 8]
```diff
@@ -72,6 +73,43 @@ export = (app: Probot) => {
     };
     await k8sApi.createNamespacedSecret(namespace,secret);
   });
+
+  //handle push event
+  app.on("push", async (context) => {
+
+    const configExists = Boolean(context.payload.commits?.reduce((acc, commit) => [...acc, ...commit.added, ...commit.modified,...commit.removed], []).find((name: String) => name == "peribolos.yaml"));
+    if(!configExists) {
+      return;
```
[the remark this answers is not in the document]
Comment (mimotej) [reply src/index.ts:83]: Resolved in latest commit.

File: src/index.ts [hunk 9]
```diff
@@ -72,6 +73,43 @@ export = (app: Probot) => {
...
+    try {
+      await k8sCustomResourceApi.createNamespacedCustomObject(
+        "tekton.dev",
+        "v1beta1",
+        namespace,
+        "taskruns",
+        payload
+      )
+    }catch(e: any){
+      app.log.error(e.message);
```
[the remark this answers is not in the document]
Comment (mimotej) [reply src/index.ts]: Resolved in latest commit.

**Review by tumido — APPROVED** [review]
/lgtm
/approve
